### PR TITLE
bpo-26515: Update C API docs to use PyModuleDef_Init()

### PR DESCRIPTION
**Pull request by berkerpeksag**
https://bugs.python.org/issue26515

### CODE REVIEW

File: Doc/extending/extending.rst [hunk 1]
```diff
@@ -204,24 +204,34 @@ usually declare a static object variable at the beginning of your file::
...
-       return m;
+       PyModule_AddObject(module, "error", SpamError);
+       return 0;
+   fail:
+       Py_XDECREF(module);
+       return -1;
    }
 
+then we need to specify the :c:func:`spam_exec` function for the
+c:var:`Py_mod_exec` slot::
```
Comment (ncoghlan) [inline Doc/extending/extending.rst:228]: Missing the leading `:` here (that appears to be what the CI error is complaining about)

File: Doc/extending/extending.rst [hunk 2]
```diff
@@ -234,8 +244,8 @@ needed to ensure that it will not be discarded, causing :c:data:`SpamError` to
 become a dangling pointer. Should it become a dangling pointer, C code which
 raises the exception could cause a core dump or other unintended side effects.
 
-We discuss the use of ``PyMODINIT_FUNC`` as a function return type later in this
-sample.
+We discuss the use of :c:macro:`PyMODINIT_FUNC` as a function return type later
```
Comment (ncoghlan) [inline Doc/extending/extending.rst:247]: With the modified example, this macro hasn't actually been referenced yet. Instead, the forward reference needs to explain that this list of slots will then be added to a `PyModuleDef` struct defining the module as a whole, which is described later.

**Review by ncoghlan — COMMENTED** [review]
This looks like a nice improvement to me. There's a forward reference that I don't think is quite right any more, and one small syntactic tweak needed to fix the build.

File: Doc/extending/extending.rst [hunk 3]
```diff
@@ -204,24 +204,34 @@ usually declare a static object variable at the beginning of your file::
 
    static PyObject *SpamError;
```
Comment (encukou) [inline Doc/extending/extending.rst:205]: This is a problem if the module is loaded more than once (for example, in multiple subinterpreters).
In that case, `spam_exec` would be executed multiple times, each time calling `PyErr_NewException` and setting `SpamError` to the newly created error object.

This pointer needs to be part of per-module state.

**Review by encukou — CHANGES_REQUESTED** [review]
Sorry for getting to this so late -- I missed the review request notification.

**Comment by vstinner** [comment]
I removed the " needs backport to 3.6" label, the 3.6 branch no longer accept bugfixes (only security fixes are accepted): https://devguide.python.org/#status-of-python-branches

**Comment by csabella** [comment]
@berkerpeksag It looks like this one was close to being merged.  When you have a chance, please review the changes requested by @ncoghlan and @encukou.  Thanks!

**Comment by JulienPalard** [comment]
@berkerpeksag, ping

**Comment by berkerpeksag** [comment]
I'll update my branch this weekend. Sorry for my late response!

**Comment by encukou** [comment]
More detailed discussion/insructions is in PEP 630.

**Comment by berkerpeksag** [comment]
@encukou so is this still needed or is PEP 630 the main reference point now?

**Comment by encukou** [comment]
Because of PEP 630's open issues, making `PyModuleDef_Init` the default way to make classes would be leading some users into a trap. I am still not comfortable making it a default in the docs, and I'd rather have users read PEP 630 and understand the benefits and tradeoffs.
Until the open issues are solved, of course.

**Comment by berkerpeksag** [comment]
Thank you. Closing this for now then.

**Pull request closed without merging**